src/main/java/spoon/support/visitor/SubInheritanceHierarchyResolver.java

```diff
@@ -0,0 +1,230 @@
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.support.visitor;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Set;
+
+import spoon.SpoonException;
+import spoon.reflect.declaration.CtClass;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.CtEnum;
+import spoon.reflect.declaration.CtPackage;
+import spoon.reflect.declaration.CtType;
+import spoon.reflect.declaration.CtTypeInformation;
+import spoon.reflect.declaration.CtTypeParameter;
+import spoon.reflect.reference.CtTypeReference;
+import spoon.reflect.visitor.Filter;
+import spoon.reflect.visitor.chain.CtConsumer;
+import spoon.reflect.visitor.chain.CtQuery;
+import spoon.reflect.visitor.chain.CtScannerListener;
+import spoon.reflect.visitor.chain.ScanningMode;
+import spoon.reflect.visitor.filter.CtScannerFunction;
+import spoon.reflect.visitor.filter.SuperInheritanceHierarchyFunction;
+import spoon.reflect.visitor.filter.TypeFilter;
+
+import static spoon.reflect.visitor.chain.ScanningMode.NORMAL;
+import static spoon.reflect.visitor.chain.ScanningMode.SKIP_ALL;
+
+/**
+ * Expects a {@link CtPackage} as input
+ * and  upon calls to forEachSubTypeInPackage produces all sub classes and sub interfaces, which extends or implements super type(s) provided in constructor and stored as `targetSuperTypes`.<br>
+ *
+ * The repeated processing of this mapping function on the same input returns only newly found sub types.
+ * The instance of {@link SubInheritanceHierarchyResolver} returns found sub types only once.
+ * So repeated call with same input package returns nothing.
+ * Create and use new instance of {@link SubInheritanceHierarchyResolver} if you need to scan the subtype hierarchy again.
+ */
+public class SubInheritanceHierarchyResolver {
+
+	CtPackage inputPackage;
+
+	private boolean includingInterfaces = true;
+	/**
+	 * Set of qualified names of all super types whose sub types we are searching for.
+	 * Each found sub type is added to this set too
+	 */
+	private Set<String> targetSuperTypes = new HashSet<>();
+	/**
+	 * if true then we have to check if type is a subtype of superClass or superInterfaces too
+	 * if false then it is enough to search in superClass hierarchy only (faster)
+	 */
+	boolean hasSuperInterface = false;
+	private boolean failOnClassNotFound = false;
+
+	public SubInheritanceHierarchyResolver(CtPackage input) {
+		inputPackage = input;

```

Comment:
What about no package at all?
The default (root) inputPackage might be taken from the factory of CtTypeInformation delivered by first call of #addSuperType.
And if somebody ever needs to search in different then root package, then we can add #setStartPackage(CtPackage) method.
That setStartPackage is better then constructor, because it can be used to recursively scan on different packages... But I personally do not need that, so I vote for removing constructor parameter